### ISSUE TITLE
The @noninjectable decorator is silently order dependent

### ISSUE THREAD

**Issue by davidparsson** [alecthomas/injector#141]
Passing (correct) example:

```
        class A:
            @injector.inject
            @injector.noninjectable('boolean')
            def __init__(self, boolean: bool = True) -> None:
                self.boolean = boolean

        i = injector.Injector()
        a = i.create_object(A)

        assert a.boolean is True
```

Failing example:
```
        class A:
            @injector.noninjectable('boolean')  # <-- @noninjectable before @inject
            @injector.inject
            def __init__(self, boolean: bool = True) -> None:
                self.boolean = boolean

        i = injector.Injector()
        a = i.create_object(A)

        assert a.boolean is True  # This will raise
```

This is dangerous, since it's an easy mistake to make and it's hard to discover.

**Comment by jstasiak** [comment]
Please provide the Python and Injector versions you're running, I can't reproduce this.

**Comment by davidparsson** [comment]
When this was reported I used Python 3.7.2 and Injector 0.15.0. so not the latest and greatest. I haven't verified that I could reproduce this again.

**Comment by jstasiak** [comment]
I can reproduce this with 0.15.0 but it's fixed in 0.16.0 and later versions, I'll add a test to make sure we don't regress.

**Comment by davidparsson** [comment]
Excellent! Thanks!
